lib/SILOptimizer/Transforms/FunctionSignatureOpts.cpp

```diff
@@ -638,8 +638,15 @@ void FunctionSignatureTransform::createFunctionSignatureOptimizedFunction() {
   SILModule &M = F->getModule();
   std::string Name = createOptimizedSILFunctionName();
   SILLinkage linkage = F->getLinkage();
+  // Specializations are similar to the original function, but slightly
+  // different: if the function is from the outside, another module may have the
+  // same specialization hence we need to be sharing, and, even if the function
+  // is public, the specialization never is: it's an implementation detail and
+  // not part of the ABI.
   if (isAvailableExternally(linkage))
     linkage = SILLinkage::Shared;
+  else if (hasPublicVisibility(linkage))
+    linkage = SILLinkage::Hidden;
```

Comment:
Oh, hm, by the same logic, wouldn't specializations of internal functions also need to be `Shared`?